### PR TITLE
Attach dangling comments to the comprehension instead of the `if` or `iter` nodes

### LINKED ISSUE #7605: Formatter instability: hard line break in binary expression
Given:

```python
[
    1 for components in  # pylint: disable=undefined-loop-variable
    b +  # integer 1 may only have decimal 01-09
    c  # negative decimal
] 
```

We format as:

```python
[
    1
    for components in b  # pylint: disable=undefined-loop-variable  # integer 1 may only have decimal 01-09
    +
    c  # negative decimal
]
```

(Notice the hard line break between `+` and `c # negative decimal`.)

Which is then formatted as:

```python
[
    1
    for components in b  # pylint: disable=undefined-loop-variable  # integer 1 may only have decimal 01-09
    + c  # negative decimal
]

```

Sourced from https://github.com/astral-sh/ruff/issues/7590 (`A_1798643049678049848.py`).

### PR DESCRIPTION
**Pull request by MichaReiser**
## Summary

This PR fixes the attachment of dangling comments for `comprehension`s. 

The comprehension formatting used to attach some dangling comments to the `iter` or `if`s node as dangling comments. 
This causes problems when the `iter` or `if` nodes have their own custom logic for handling dangling comments. 

For example, binary expressions use the dangling comments to store comments after the operator. 
Attaching dangling `iter` comments to the binary expression misguides the binary expression formatting because
it now formats it as if it has trailing operator comments, which it has not. 

This PR fixes the issue by attaching all special comprehension comments as dangling comments to the comprehension and use indexing to split the comments during formatting.

Closes #7605 
Closes https://github.com/astral-sh/ruff/issues/6588

## Test Plan

`cargo test`. Added test, added debug assertions for nodes that can never have dangling comments (wasn't possible before)

### CODE REVIEW

**Comment by MichaReiser** [comment]
Current dependencies on/for this PR:
* main
  * **PR #7693** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/7693" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/astral-sh/ruff/7693?utm_source=stack-comment).

**Review by konstin — APPROVED** [review]

File: crates/ruff_python_formatter/src/module/mod_module.rs [hunk 1]
```diff
@@ -37,10 +37,11 @@ impl FormatNodeRule<ModModule> for FormatModModule {
 
     fn fmt_dangling_comments(
         &self,
-        _dangling_comments: &[SourceComment],
+        dangling_comments: &[SourceComment],
         _f: &mut PyFormatter,
     ) -> FormatResult<()> {
-        // Handled as part of `fmt_fields`
+        // Node can't have dangling comments.
+        debug_assert!(dangling_comments.is_empty());
```
Comment (charliermarsh) [inline crates/ruff_python_formatter/src/module/mod_module.rs:44]: Not: I think these changes to fmt_dangling_comments could be a separate PR.

File: crates/ruff_python_formatter/Cargo.toml [hunk 2]
```diff
@@ -38,7 +38,7 @@ tracing = { workspace = true }
 unicode-width = { workspace = true }
 
 [dev-dependencies]
-ruff_formatter = { path = "../ruff_formatter", features = ["serde"] }
+ruff_formatter = { path = "../ruff_formatter" }
```
Comment (charliermarsh) [inline crates/ruff_python_formatter/Cargo.toml:41]: What’s this?

**Review by charliermarsh — APPROVED** [review]

File: crates/ruff_python_formatter/src/module/mod_module.rs (hunk 1 above)
Comment (MichaReiser) [reply crates/ruff_python_formatter/src/module/mod_module.rs:44]: Yeah, that's fair. I used it to verify that no other node does something similar. But it could certainly have been its own PR. Would you prefer if I split it out?

File: crates/ruff_python_formatter/Cargo.toml (hunk 2 above)
Comment (MichaReiser) [reply crates/ruff_python_formatter/Cargo.toml:41]: A fix-up from my changes. Could have been it's own PR too, I guess. There's a `serde` feature. The `ruff_formatter` serde feature should depend on whether `serde` is enabled for `ruff_python_formatter`

File: crates/ruff_python_formatter/src/module/mod_module.rs (hunk 1 above)
Comment (charliermarsh) [reply crates/ruff_python_formatter/src/module/mod_module.rs:44]: I think it's fine as-is actually -- I assume these debug checks _don't_ pass without these changes.

Comment (MichaReiser) [reply crates/ruff_python_formatter/src/module/mod_module.rs:44]: Yeah they don't. That's how I discovered that comprehension attach dangling comments to other nodes in the fist place (because I wanted to add the assertions and then got really confused why some of the nodes suddenly had dangling comments)

Comment (charliermarsh) [reply crates/ruff_python_formatter/src/module/mod_module.rs:44]: Totally fine with merging as-is.
